### PR TITLE
Rel 1.7 mongoose cluster

### PR DESCRIPTION
**Pull request by ludwikbukowski**
This PR addresses https://github.com/esl/MongooseIM/issues/685

Proposed changes include:
* Replacing commands `add_to_cluster` and `remove_from_cluster` with brand-new
`join_cluster` and `leave_cluster` in _ejabberd_admin_ module. The commands use module _mongoose_cluster_ for the cluster management hence it's done completely at the Erlang level
* updated tests in _cluster_commands_SUITE_

### CODE REVIEW

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 1]
```diff
@@ -44,7 +44,8 @@
          dump_to_textfile/1, dump_to_textfile/2,
          mnesia_change_nodename/4,
          restore/1, % Still used by some modules%%
-         get_loglevel/0]).
+         get_loglevel/0,
+         maybe_join_cluster/1, maybe_leave_cluster/0]).
```
Comment (erszcz) [inline apps/ejabberd/src/ejabberd_admin.erl]: Should we export `maybe_*` functions as public API? In case of public interface I'd be for making the module API calls shorter. If after doing the checks (the "maybe"s hint at) the actual implementation needs to be put into deeper functions, we can call them `do_*` (`do_join_cluster`...).

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 2]
```diff
@@ -147,7 +148,17 @@ commands() ->
      #ejabberd_commands{name = reload_cluster, tags = [server],
                         desc = "Reload configuration file in the cluster",
                         module = ejabberd_config, function = reload_cluster,
-                        args = [], result = {res, restuple}}
+                        args = [], result = {res, restuple}},
+      #ejabberd_commands{name = join_cluster, tags = [server],
```
Comment (erszcz) [inline apps/ejabberd/src/ejabberd_admin.erl]: Indentation is inconsistent with the former command (line 147/148). Tabs vs spaces issue? BTW, in the following lines - why 2 spaces, when everything around is 4 spaces?

Comment (ludwikbukowski) [inline apps/ejabberd/src/ejabberd_admin.erl]: it's an "Easter egg"

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 3]
```diff
@@ -169,7 +169,7 @@ commands() ->
 -spec maybe_join_cluster(string()) -> {ok, string()} | {pang, string()} | {mnesia_error, string()} | {error, string()}.
 maybe_join_cluster(NodeString) ->
   NodeAtom = list_to_atom(NodeString),
-  NodeList = nodes(),
+  NodeList = mnesia:system_info(running_db_nodes),
   case lists:any(fun(El) -> El == NodeAtom end, NodeList) of
```
Comment (erszcz) [inline apps/ejabberd/src/ejabberd_admin.erl]: `lists:member`?

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 4]
```diff
@@ -197,7 +197,7 @@ join_cluster(Node) ->
 
 -spec maybe_leave_cluster() -> {ok, string()} | {error, term()} | {not_in_cluster, string()}.
 maybe_leave_cluster() ->
-  NodeList = nodes(),
+  NodeList = mnesia:system_info(running_db_nodes),
```
Comment (erszcz) [inline apps/ejabberd/src/ejabberd_admin.erl]: I'm not sure this can happen or not, but what if due to some inconsistency (network issue) we're not in `running_db_nodes`, but are in `db_nodes`? Again, maybe it's impossible, but some food for thought.

File: rel/files/mongooseimctl [hunk 5]
```diff
@@ -70,8 +90,8 @@ help ()
     echo "  debug  Attach an interactive Erlang shell to a running MongooseIM node"
     echo "  live   Start MongooseIM node in live (interactive) mode"
     echo "MongooseIM cluster management commads:"
-    echo "  add_to_cluster other_node_name       Add current node to cluster"
-    echo "  remove_from_cluster dead_node_name   Remove other node from cluster"
+    echo "  join_cluster other_node_name       Add current node to cluster"
+    echo "  leave_cluster   Leave current node from the cluster"
```
Comment (erszcz) [inline rel/files/mongooseimctl]: Which looks better?

```
$ git --help
usage: git [--version] [--help] [-C <path>] [-c name=value]
...

The most commonly used git commands are:
   add        Add file contents to the index
   bisect     Find by binary search the change that introduced a bug
   branch     List, create, or delete branches
   checkout   Checkout a branch or paths to the working tree
   clone      Clone a repository into a new directory
   commit     Record changes to the repository
   diff       Show changes between commits, commit and working tree, etc
   fetch      Download objects and refs from another repository
```

Or?

```
$ git --help
usage: git [--version] [--help] [-C <path>] [-c name=value]
...

The most commonly used git commands are:
   add  Add file contents to the index
   bisect  Find by binary search the change that introduced a bug
   branch  List, create, or delete branches
   checkout  Checkout a branch or paths to the working tree
   clone  Clone a repository into a new directory
   commit  Record changes to the repository
   diff  Show changes between commits, commit and working tree, etc
   fetch  Download objects and refs from another repository
```

File: test/ejabberd_tests/tests/cluster_commands_SUITE.erl [hunk 6]
```diff
@@ -79,18 +77,38 @@ init_per_group(Group, Config) when Group == clustered orelse Group == ejabberdct
             remove_node_from_cluster(Config1),
             {skip, nondistributed_sm}
     end;
+
+init_per_group(clustering, _Config) ->
+    case is_sm_distributed() of
+        true ->
+            % escalus:create_users(Config, escalus:get_users([alice, clusterguy]));
```
Comment (ludwikbukowski) [inline test/ejabberd_tests/tests/cluster_commands_SUITE.erl]: Ops. I will remove this line before merge

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 7]
```diff
@@ -155,6 +166,59 @@ commands() ->
...
+            {ok, String};
+        _ ->
+            do_join_cluster(NodeAtom)
+    end.
+
+do_join_cluster(Node) ->
+    try mongoose_cluster:join(Node) of
+        ok ->
+            String = io_lib:format("You have successfully joined the node ~p to
+      the cluster with node member ~p~n", [node(), Node]),
```
Comment (michalwski) [inline apps/ejabberd/src/ejabberd_admin.erl]: Is it intentional to add a new line here? If so I'd rather add `~n` and not split the string into many lines. The indentation is unintentional here.

**Comment by michalwski** [comment]
I'd invest some more time in testing the negative test cases like node tries to join twice or node is removed twice. Basically consult coveralls.io [1] for your build and see what new lines added by you are not covered yet.

[1]. https://coveralls.io/builds/5617628

File: apps/ejabberd/src/ejabberd_admin.erl [hunk 8]
```diff
@@ -155,6 +166,59 @@ commands() ->
...
+        E:R ->
+            {error, {E, R}}
+    end.
+
+-spec leave_cluster() -> {ok, string()} | {error, term()} | {not_in_cluster, string()}.
+leave_cluster() ->
+    NodeList = mnesia:system_info(running_db_nodes),
+    ThisNode = node(),
+    case NodeList of
+        [ThisNode] ->
```
Comment (michalwski) [inline apps/ejabberd/src/ejabberd_admin.erl]: This matching doesn't look good to me. What will happen if we have more then 1 node in running_db_nodes? A test for that would be more than welcome!

File: rel/files/mongooseimctl [hunk 9]
```diff
@@ -219,8 +239,8 @@ done
 
 case $1 in
     'start') start;;
-    'add_to_cluster') add_to_cluster;;
-    'remove_from_cluster') remove_from_cluster;;
+    'join_cluster') join_cluster;;
```
Comment (michalwski) [inline rel/files/mongooseimctl]: Why this name change? I understand its better but I see no updated in the doc about this change.

File: rel/files/mongooseimctl [hunk 10]
```diff
@@ -219,8 +239,8 @@ done
...
+    'leave_cluster') leave_cluster;;
```
Comment (michalwski) [inline rel/files/mongooseimctl]: As far as I understand leave_cluster should be run remove a node (different one) from a cluster. If we have 3 nodes A, B, C and we want to remove C from the cluster. Then we need to run this command on A or B. Am I right? If so the `leave_cluster` name is not intuitive to me.

File: apps/ejabberd/src/ejabberd_admin.erl (hunk 8 above)
Comment (ludwikbukowski) [inline apps/ejabberd/src/ejabberd_admin.erl]: If there is more than one node the case would match here [1] and it would mean that node is in any cluster and can leave it. Otherwise [2] there is no cluster to leave because `mnesia:system_info(running_db_nodes)` returned just the local node
**However you're right about the lacking tests!**

[1]https://github.com/esl/MongooseIM/blob/rel-1.7-mongoose-cluster/apps/ejabberd/src/ejabberd_admin.erl#L206
[2]https://github.com/esl/MongooseIM/blob/rel-1.7-mongoose-cluster/apps/ejabberd/src/ejabberd_admin.erl#L203

File: rel/files/mongooseimctl (hunk 9 above)
Comment (ludwikbukowski) [inline rel/files/mongooseimctl]: Ok, I will update the docs

File: rel/files/mongooseimctl (hunk 10 above)
Comment (ludwikbukowski) [inline rel/files/mongooseimctl]: Command `join_cluster` takes as a parameter some cluster member and it's called from joining node ("the alone one")
Command 'leave_cluster' takes no parameters and it's called from  the node which is going to leave the cluster.
In your scenario, if we have clustered nodes A, B and C and we want to remove C we have to call `leave_cluster` from node C
What I've done in this PR was just a pure command line interface for this module [1] and maybe I didn't understand the task correctly?
[1]https://github.com/esl/MongooseIM/blob/20d40e87ab492013b77c32153fbce392890efae1/apps/ejabberd/src/mongoose_cluster.erl

File: apps/ejabberd/src/ejabberd_admin.erl (hunk 7 above)
Comment (ludwikbukowski) [inline apps/ejabberd/src/ejabberd_admin.erl]: I must have split it because the line was too long but not having in mind that it's text to display!
I will fix it

File: rel/files/mongooseimctl (hunk 10 above)
Comment (michalwski) [inline rel/files/mongooseimctl]: Thanks for the explanation. Now I understand :)

File: apps/ejabberd/src/ejabberd_admin.erl (hunk 8 above)
Comment (michalwski) [inline apps/ejabberd/src/ejabberd_admin.erl]: Right. It was before my morning coffee. Thanks for explaining anyway :)

File: test/ejabberd_tests/tests/cluster_commands_SUITE.erl [hunk 11]
```diff
@@ -134,3 +152,53 @@ set_master_test(ConfigIn) ->
     ejabberdctl("set_master", ["self"], ConfigIn),
     [MasterNode] = rpc_call(mnesia, table_info, [TableName, master_nodes]).
 
+join_successful(Config) ->
+    Node2 = ct:get_config(ejabberd2_node),
```
Comment (ludwikbukowski) [inline test/ejabberd_tests/tests/cluster_commands_SUITE.erl]: I'm going also to stop using config values like `ejabberd2_node` and `ejabberd_node` and convert suite to use `hosts` block defined here [1]

[1]https://github.com/esl/MongooseIM/blob/master/test/ejabberd_tests/test.config#L26

File: test/ejabberd_tests/tests/cluster_commands_SUITE.erl [hunk 12]
```diff
@@ -17,60 +17,86 @@
...
+        {group, ejabberdctl},
+        {group, clustering_two},
+        {group, clustering_theree}].
 groups() ->
     [{clustered, [], [one_to_one_message]},
-     {ejabberdctl, [], [set_master_test]}].
-
+        {clustering_two, [],
+            [join_successful,
+            leave_successful,
```
Comment (michalwski) [inline test/ejabberd_tests/tests/cluster_commands_SUITE.erl]: Indentation :) please align the whole list to line 42.

File: test/ejabberd_tests/tests/cluster_commands_SUITE.erl [hunk 13]
```diff
@@ -17,60 +17,86 @@
...
+            [cluster_of_theree,
+             leave_the_theree]},
+        {ejabberdctl, [], [set_master_test]}].
 suite() ->
+    require_all_nodes() ++
     escalus:suite().
 
+require_all_nodes() ->
+    [{require, mim_node, {hosts, mim, node}},
+        {require, mim_node2, {hosts, mim2, node}},
```
Comment (michalwski) [inline test/ejabberd_tests/tests/cluster_commands_SUITE.erl]: Indentation again :)

File: test/ejabberd_tests/tests/cluster_commands_SUITE.erl [hunk 14]
```diff
@@ -103,20 +158,20 @@ one_to_one_message(ConfigIn) ->
...
+        Stanza1 = escalus:wait_for_stanza(ClusterGuy, 5000),
+        escalus:assert(is_chat_message, [<<"Hi!">>], Stanza1),
+
+        %% When ClusterGuy sends a response
+        Msg2 = escalus_stanza:chat_to(Alice, <<"Oh hi!">>),
+        escalus:send(ClusterGuy, Msg2),
+        %% Then Alice also receives it
+        Stanza2 = escalus:wait_for_stanza(Alice, 5000),
+        escalus:assert(is_chat_message, [<<"Oh hi!">>], Stanza2)
+                                                         end).
```
Comment (michalwski) [inline test/ejabberd_tests/tests/cluster_commands_SUITE.erl]: It'd be more readable to have `end` at the same level as `escalus:story`.

File: test/ejabberd_tests/tests/distributed_helper.erl [hunk 15]
```diff
@@ -43,42 +42,58 @@ wait_until_started(Cmd, Retries) ->
...
     escalus_ct:rpc_call(Node, M, F, A, TimeOut, Cookie).
+
+get_or_fail(Key) ->
+    Val = ct:get_config(Key),
+    Val == undefined andalso error({undefined, Key}),
+    Val.
+
+
+%% MongooseIM node names
+mim() ->
```
Comment (michalwski) [inline test/ejabberd_tests/tests/distributed_helper.erl]: I'd move this function  (also mim2 and fed) to `ejabberd_node_utils`. IMO they better suite there. Ideally would be to rename the `ejabberd_node_utils` to  `mongoose_node_utils`, or just `node_utils`. But this may introduce too many changes for this PR so maybe it's a subject of a separate PR.
